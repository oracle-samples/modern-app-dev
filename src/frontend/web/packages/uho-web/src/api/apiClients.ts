/*
 * Copyright (c) 2023 Oracle and/or its affiliates.
 * Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracl.com/licenses/upl/
 */
export { AppointmentApi } from '@uho/appointment-api-client';
export { EncounterApi } from '@uho/encounter-api-client';
export { FrontendApi } from './frontendApi';
export { PatientApi } from '@uho/patient-api-client/dist/api-client';
export { ProviderApi } from '@uho/provider-api-client/dist/api-client';
