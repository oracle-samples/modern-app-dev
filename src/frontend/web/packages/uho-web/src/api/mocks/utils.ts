/*
 * Copyright (c) 2023 Oracle and/or its affiliates.
 * Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracl.com/licenses/upl/
 */

type Options = {
  delay?: number;
  error?: Error;
};
const defaultDelay = 0;

export function mock<T>(data: T, options?: Options): Promise<T> {
  return new Promise<T>((resolve, reject) => {
    setTimeout(() => {
      options?.error ? reject(options.error) : resolve(data);
    }, options?.delay || defaultDelay);
  });
}
export function reject<T>(options?: Options): Promise<T> {
  return new Promise<T>((_, reject) => {
    setTimeout(() => {
      reject(options?.error);
    }, options?.delay || defaultDelay);
  });
}

export function mockDate() {
  return new Date(1654733487565);
}
