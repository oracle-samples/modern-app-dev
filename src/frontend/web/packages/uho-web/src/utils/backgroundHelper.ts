/*
 * Copyright (c) 2023 Oracle and/or its affiliates.
 * Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracl.com/licenses/upl/
 */
import { ComponentProps } from 'preact/compat';
import { Avatar } from 'oj-c/avatar';

type AvatarProps = ComponentProps<typeof Avatar>;
const backgrounds: AvatarProps['background'][] = [
  'neutral',
  'orange',
  'green',
  'teal',
  'blue',
  'slate',
  'pink',
  'purple',
  'lilac',
  'gray'
];

// simple hash function to generate a number from 0 - brackets
function computeHash(key: string, brackets: number) {
  const hash = key.length + key.charCodeAt(0);
  return hash % brackets || 0;
}

export function getBackgroundByName(name: string) {
  return backgrounds[computeHash(name, backgrounds.length)];
}
