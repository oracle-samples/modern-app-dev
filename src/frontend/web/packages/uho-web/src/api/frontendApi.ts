/*
 * Copyright (c) 2023 Oracle and/or its affiliates.
 * Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracl.com/licenses/upl/
 */
import { Patient } from '@uho/patient-api-client/dist/api-client';
import { Provider } from '@uho/provider-api-client/dist/api-client';
import { UserRole } from '../utils/authProvider';

export type Apm = {
  serviceName: string;
  webApplication: string;
  ociDataUploadEndpoint: string;
  publicDataKey: string;
};

export type User = (Patient | Provider) & { role: UserRole };

export class Configuration {
  basePath: string;
  constructor(config: { basePath: string }) {
    this.basePath = config.basePath;
  }
}

export class FrontendApi {
  public constructor(public configuration: Configuration) {}

  public getUserInformation(): Promise<User> {
    return fetch(`${this.configuration.basePath}/userInformation`, {
      method: 'GET',
      headers: {
        accept: '*/*'
      }
    })
      .then((result) => result.json())
      .then((users) => users[0]);
  }

  public getApmInformation(): Promise<Apm> {
    return fetch(`${this.configuration.basePath}/apmInformation`, {
      method: 'GET',
      headers: {
        accept: '*/*'
      }
    }).then((result) => result.json());
  }

  public login(role: UserRole) {
    window.location.href = `oauth/login/${role.toLowerCase()}`;
  }

  public signout() {
    window.location.href = '/home/oauth/logout/';
  }
}
