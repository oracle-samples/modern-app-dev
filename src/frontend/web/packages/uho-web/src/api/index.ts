/*
 * Copyright (c) 2023 Oracle and/or its affiliates.
 * Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracl.com/licenses/upl/
 */
import { PatientApi, EncounterApi, AppointmentApi, ProviderApi, FrontendApi } from './apiClients';
import { Configuration as FrontendConfiguration } from './frontendApi';
import { Configuration as ProviderConfiguration } from '@uho/provider-api-client/dist/api-client';
import { Configuration as EncounterConfiguration, EncounterApiInterface } from '@uho/encounter-api-client';
import { Configuration as AppointmentConfiguration, AppointmentApiInterface } from '@uho/appointment-api-client';
import { Configuration as PatientConfiguration, PatientApiInterface } from '@uho/patient-api-client/dist/api-client';

const basePath = '/home/api';

export const providerApi = new ProviderApi(
  new ProviderConfiguration({
    basePath
  })
);
export const patientApi: PatientApiInterface = new PatientApi(new PatientConfiguration({ basePath }));
export const encounterApi: EncounterApiInterface = new EncounterApi(new EncounterConfiguration({ basePath }));
export const appointmentApi: AppointmentApiInterface = new AppointmentApi(new AppointmentConfiguration({ basePath }));
export const frontendApi: FrontendApi = new FrontendApi(new FrontendConfiguration({ basePath }));
