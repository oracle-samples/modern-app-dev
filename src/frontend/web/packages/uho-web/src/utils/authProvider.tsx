/*
 * Copyright (c) 2023 Oracle and/or its affiliates.
 * Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracl.com/licenses/upl/
 */
import { h, ComponentChildren, createContext } from 'preact';
import { useCallback, useContext } from 'preact/hooks';
import { RouteProps } from 'react-router';
import { Navigate } from 'react-router-dom';
import { Loading } from 'components/loading';
import { useLocalStorage } from 'hooks/useLocalStorage';
import { frontendApi } from 'api';
import { User } from 'api/frontendApi';
import { useQuery } from '@tanstack/react-query';
import { warn } from 'ojs/ojlogger';

export type UserRole = 'PATIENT' | 'PROVIDER';

type Context = {
  user: User | null;
  signin: () => Promise<User | null>;
  signout: () => void;
};

function useProvideAuth(): Context {
  const [user, setUser] = useLocalStorage<User | null>('user');

  const signin = useCallback(async () => {
    let userInfo: User | null = null;
    try {
      userInfo = await frontendApi.getUserInformation();
    } catch {
      warn('No active user session found.');
    }
    setUser(userInfo);
    return userInfo;
  }, []);

  const signout = useCallback(async () => {
    setUser(null);
    return frontendApi.signout();
  }, []);

  return {
    user,
    signin,
    signout
  };
}

// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
export const AuthContext = createContext<Context>(undefined!);

export function useAuth() {
  return useContext(AuthContext);
}

export function AuthProvider({ children }: { children: ComponentChildren }) {
  const auth = useProvideAuth();
  return <AuthContext.Provider value={auth}>{children}</AuthContext.Provider>;
}

export function RequireAuth({ children, role }: { children?: ComponentChildren; role: UserRole } & RouteProps) {
  const { signin } = useAuth();
  const { data, isLoading, isError } = useQuery<User | null, void>(['signin'], signin);

  if (isLoading) {
    return <Loading />;
  }
  if (!data || isError) {
    return <Navigate to="/" />;
  }
  if (data.role !== role) {
    return <Navigate to="/404" />;
  }
  return children;
}
