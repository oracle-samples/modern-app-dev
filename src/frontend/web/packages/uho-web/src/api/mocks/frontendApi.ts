/*
 * Copyright (c) 2023 Oracle and/or its affiliates.
 * Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracl.com/licenses/upl/
 */
import { UserRole } from '../../utils/authProvider';
import { Apm, Configuration, FrontendApi, User } from '../frontendApi';
import { mock } from './utils';

/**
 * Mocked Frontend API for local development
 */
export class MockFrontendApi extends FrontendApi {
  public constructor(configuration: Configuration) {
    super(configuration);
  }

  public getUserInformation(): Promise<User> {
    const userString = localStorage.getItem('user');
    if (userString) {
      return Promise.resolve(JSON.parse(userString));
    }
    return Promise.reject();
  }

  public getApmInformation(): Promise<Apm> {
    return mock({
      publicDataKey: 'publicDataKey',
      ociDataUploadEndpoint: 'ociDataUploadEndpoint',
      serviceName: 'serviceName',
      webApplication: 'webApplication'
    });
  }

  public async login(role: UserRole): Promise<void> {
    const user = {
      role,
      name: role === 'PATIENT' ? 'Patient XYZ' : 'Provider XYZ',
      username: role === 'PATIENT' ? 'DevUserId' : 'jsmith',
      email: role === 'PATIENT' ? 'patient@oracle.com' : 'provider@oracle.com'
    };
    localStorage.setItem('user', JSON.stringify(user));
    window.location.reload();
  }

  public signout() {
    window.location.href = '/home';
  }
}
