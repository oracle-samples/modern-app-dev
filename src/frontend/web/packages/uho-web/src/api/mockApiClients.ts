/*
 * Copyright (c) 2023 Oracle and/or its affiliates.
 * Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracl.com/licenses/upl/
 */
import { MockAppointmentApi as AppointmentApi } from './mocks/appointmentApi';
import { MockEncounterApi as EncounterApi } from './mocks/encounterApi';
import { MockFrontendApi as FrontendApi } from './mocks/frontendApi';
import { MockPatientApi as PatientApi } from './mocks/patientApi';
import { MockProviderApi as ProviderApi } from './mocks/providerApi';

export { AppointmentApi, EncounterApi, FrontendApi, PatientApi, ProviderApi };
