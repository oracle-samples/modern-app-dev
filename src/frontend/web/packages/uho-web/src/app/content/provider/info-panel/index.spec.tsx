/*
 * Copyright (c) 2023 Oracle and/or its affiliates.
 * Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracl.com/licenses/upl/
 */
import { InfoPanel } from '.';
import { h } from 'preact';
import { snapshot } from '../../../../../test/jetHelper';
import { MemoryRouter } from 'react-router-dom';
import { AuthContext } from '../../../../utils/authProvider';
import { User } from 'api/frontendApi';

const user: User = {
  role: 'PROVIDER',
  username: 'jsmith'
};

snapshot(
  'Provider: InfoPanel',
  <AuthContext.Provider value={{ user, signin: () => Promise.resolve(user), signout: () => console.log('signout') }}>
    <MemoryRouter>
      <InfoPanel />
    </MemoryRouter>
  </AuthContext.Provider>
);
