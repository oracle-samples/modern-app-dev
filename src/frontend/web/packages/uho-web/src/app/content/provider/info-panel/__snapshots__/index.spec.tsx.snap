// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot tests for Provider: InfoPanel Snapshot: Component - Provider: InfoPanel 1`] = `
<DocumentFragment>
  <div
    data-oj-binding-provider="preact"
  >
    <div
      class="oj-panel oj-panel-shadow-sm"
    >
      <div
        class="oj-sm-margin-3x"
      >
        <div
          class="provider-profile"
        >
          <div
            class="oj-flex-bar oj-sm-margin-8x-bottom"
          >
            <span
              class="oj-flex-bar-start oj-badge oj-badge-subtle oj-badge-info"
            >
              Your Profile
            </span>
          </div>
          <div
            class="person oj-flex oj-sm-flex-direction-column oj-sm-margin-4x-bottom"
          >
            <oj-c-avatar
              class="oj-flex-item oj-sm-margin-5x-bottom oj-complete"
              role="img"
            >
              <div
                aria-label="Avatar of FirstName + LastName"
                class="x4exih oj-c-avatar-bg-pink oj-c-avatar-2xl"
                role="img"
              >
                <div
                  aria-hidden="true"
                  class="k1vxov _1fma8t6 _1lhcr86 u5bbjp"
                />
              </div>
            </oj-c-avatar>
            <span
              class="oj-typography-heading-md oj-flex-item"
            >
               
            </span>
            <span
              class="oj-typography-semi-bold oj-typography-body-md oj-text-color-secondary oj-flex-item"
            />
            <div
              class="oj-flex-item"
            >
              <oj-picto-chart
                class="oj-dvtbase oj-pictochart oj-component-initnode oj-complete"
              >
                <div
                  class="oj-helper-detect-contraction"
                >
                  <div
                    style="width: 200%; height: 200%;"
                  />
                </div>
                <div
                  class="oj-helper-detect-expansion"
                >
                  <div />
                </div>
                <template
                  slot="itemTemplate"
                />
                <div
                  style="visibility: hidden;"
                />
                <svg
                  height="100%"
                  width="100%"
                >
                  <defs />
                  <g
                    cursor="default"
                    font-family="-apple-system,BlinkMacSystemFont,'Segoe UI','Helvetica Neue',Arial,sans-serif"
                    font-size="12px"
                    font-weight="400"
                  >
                    <g />
                  </g>
                </svg>
              </oj-picto-chart>
            </div>
          </div>
          <div
            class="contact"
          >
            <oj-form-layout
              class="oj-form-layout oj-complete oj-formlayout-max-cols-1"
            >
              <div
                class="oj-form oj-enabled oj-efficient oj-form-cols oj-formlayout-form-across"
                data-oj-context=""
                data-oj-formlayout-bonus-dom=""
                data-oj-internal=""
                style="column-count: 1; webkit-column-count: 1;"
              >
                <div
                  class="oj-helper-detect-contraction"
                >
                  <div
                    style="width: 200%; height: 200%;"
                  />
                </div>
                <div
                  class="oj-helper-detect-expansion"
                >
                  <div />
                </div>
                <div
                  class="oj-flex"
                  data-oj-formlayout-bonus-dom=""
                  data-oj-internal=""
                >
                  <div
                    class="oj-flex-item oj-formlayout-no-label-flex-item oj-formlayout-nested-labelvalue"
                    data-oj-formlayout-bonus-dom=""
                    data-oj-internal=""
                    style="flex-grow: 1; flex-shrink: 1; max-width: calc(((100% / 1) - ((var(--oj-form-layout-column-gutter) * (0) / 1))));"
                  >
                    <oj-label-value
                      class="oj-label-value oj-form oj-form-control-label-inside oj-formlayout-form-across oj-complete"
                      data-oj-colspan="1"
                      data-oj-column-gap="var(--oj-form-layout-column-gutter)"
                    >
                      <div
                        class="oj-flex"
                        data-oj-context=""
                        data-oj-internal=""
                      >
                        <div
                          class="oj-flex-item"
                          data-oj-context=""
                          data-oj-internal=""
                          style="flex-grow: 0; flex-shrink: 1; flex-basis: 100%; width: 100%; max-width: 100%;"
                        >
                          <oj-label
                            class="oj-label oj-component oj-complete"
                            label-id="f20"
                            slot="label"
                          >
                            <div
                              class="oj-label-group"
                            >
                              <label
                                class="oj-component-initnode"
                                data-oj-internal=""
                              >
                                Email
                              </label>
                            </div>
                          </oj-label>
                        </div>
                        <div
                          class="oj-flex-item"
                          data-oj-context=""
                          data-oj-internal=""
                          style="flex: 1 1 0px;"
                        >
                          <a
                            slot="value"
                          />
                        </div>
                      </div>
                    </oj-label-value>
                  </div>
                </div>
                <div
                  class="oj-flex"
                  data-oj-formlayout-bonus-dom=""
                  data-oj-internal=""
                >
                  <div
                    class="oj-flex-item oj-formlayout-no-label-flex-item oj-formlayout-nested-labelvalue"
                    data-oj-formlayout-bonus-dom=""
                    data-oj-internal=""
                    style="flex-grow: 1; flex-shrink: 1; max-width: calc(((100% / 1) - ((var(--oj-form-layout-column-gutter) * (0) / 1))));"
                  >
                    <oj-label-value
                      class="oj-label-value oj-form oj-form-control-label-inside oj-formlayout-form-across oj-complete"
                      data-oj-colspan="1"
                      data-oj-column-gap="var(--oj-form-layout-column-gutter)"
                    >
                      <div
                        class="oj-flex"
                        data-oj-context=""
                        data-oj-internal=""
                      >
                        <div
                          class="oj-flex-item"
                          data-oj-context=""
                          data-oj-internal=""
                          style="flex-grow: 0; flex-shrink: 1; flex-basis: 100%; width: 100%; max-width: 100%;"
                        >
                          <oj-label
                            class="oj-label oj-component oj-complete"
                            label-id="f20"
                            slot="label"
                          >
                            <div
                              class="oj-label-group"
                            >
                              <label
                                class="oj-component-initnode"
                                data-oj-internal=""
                              >
                                Phone
                              </label>
                            </div>
                          </oj-label>
                        </div>
                        <div
                          class="oj-flex-item"
                          data-oj-context=""
                          data-oj-internal=""
                          style="flex: 1 1 0px;"
                        >
                          <a
                            slot="value"
                          />
                        </div>
                      </div>
                    </oj-label-value>
                  </div>
                </div>
                <div
                  class="oj-flex"
                  data-oj-formlayout-bonus-dom=""
                  data-oj-internal=""
                >
                  <div
                    class="oj-flex-item oj-formlayout-no-label-flex-item oj-formlayout-nested-labelvalue"
                    data-oj-formlayout-bonus-dom=""
                    data-oj-internal=""
                    style="flex-grow: 1; flex-shrink: 1; max-width: calc(((100% / 1) - ((var(--oj-form-layout-column-gutter) * (0) / 1))));"
                  >
                    <oj-label-value
                      class="oj-label-value oj-form oj-form-control-label-inside oj-formlayout-form-across oj-complete"
                      data-oj-colspan="1"
                      data-oj-column-gap="var(--oj-form-layout-column-gutter)"
                    >
                      <div
                        class="oj-flex"
                        data-oj-context=""
                        data-oj-internal=""
                      >
                        <div
                          class="oj-flex-item"
                          data-oj-context=""
                          data-oj-internal=""
                          style="flex-grow: 0; flex-shrink: 1; flex-basis: 100%; width: 100%; max-width: 100%;"
                        >
                          <oj-label
                            class="oj-label oj-component oj-complete"
                            label-id="f20"
                            slot="label"
                          >
                            <div
                              class="oj-label-group"
                            >
                              <label
                                class="oj-component-initnode"
                                data-oj-internal=""
                              >
                                Facility
                              </label>
                            </div>
                          </oj-label>
                        </div>
                        <div
                          class="oj-flex-item"
                          data-oj-context=""
                          data-oj-internal=""
                          style="flex: 1 1 0px;"
                        >
                          <span
                            class="oj-text-color-primary"
                            slot="value"
                          />
                        </div>
                      </div>
                    </oj-label-value>
                  </div>
                </div>
                <div
                  class="oj-flex"
                  data-oj-formlayout-bonus-dom=""
                  data-oj-internal=""
                >
                  <div
                    class="oj-flex-item oj-formlayout-no-label-flex-item oj-formlayout-nested-labelvalue"
                    data-oj-formlayout-bonus-dom=""
                    data-oj-internal=""
                    style="flex-grow: 1; flex-shrink: 1; max-width: calc(((100% / 1) - ((var(--oj-form-layout-column-gutter) * (0) / 1))));"
                  >
                    <oj-label-value
                      class="oj-label-value oj-form oj-form-control-label-inside oj-formlayout-form-across oj-complete"
                      data-oj-colspan="1"
                      data-oj-column-gap="var(--oj-form-layout-column-gutter)"
                    >
                      <div
                        class="oj-flex"
                        data-oj-context=""
                        data-oj-internal=""
                      >
                        <div
                          class="oj-flex-item"
                          data-oj-context=""
                          data-oj-internal=""
                          style="flex-grow: 0; flex-shrink: 1; flex-basis: 100%; width: 100%; max-width: 100%;"
                        >
                          <oj-label
                            class="oj-label oj-component oj-complete"
                            label-id="f20"
                            slot="label"
                          >
                            <div
                              class="oj-label-group"
                            >
                              <label
                                class="oj-component-initnode"
                                data-oj-internal=""
                              >
                                Address
                              </label>
                            </div>
                          </oj-label>
                        </div>
                        <div
                          class="oj-flex-item"
                          data-oj-context=""
                          data-oj-internal=""
                          style="flex: 1 1 0px;"
                        >
                          <span
                            class="oj-text-color-primary"
                            slot="value"
                          />
                        </div>
                      </div>
                    </oj-label-value>
                  </div>
                </div>
                <div
                  data-oj-formlayout-workaround-dom=""
                  style="display: table; visibility: hidden;"
                />
              </div>
            </oj-form-layout>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    id="__oj_layerhost_container"
  />
</DocumentFragment>
`;
